File: countdown_api/countdown_price_retriever.py
import re
from datetime import datetime
import pytz
import requests
from price_definition.price import ProductPriceModel
from retrying import retry


class CountdownPriceRetriever:

    @staticmethod
    @retry(stop_max_attempt_number=20, wait_random_min=10000, wait_random_max=60000)
    def request_product_price(store_product_code: str):
        """
        Retrieve product price from countdown api for provided product code and return json object of response. Retry
        request after a period of time if request fails.
        """

        url = f'https://shop.countdown.co.nz/api/v1/products/{store_product_code}'

        cookies = {"cw-lrkswrdjp": "dm-Pickup,f-9036,a-168,s-38"}

        headers = {
            # pretend to be chrome
            'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) '
                          'Chrome/90.0.4430.93 Safari/537.36',
            'Accept-Encoding': 'gzip, deflate, br',
            'Accept': 'application/json',
            'Content-Type': 'application/json',
            'X-Requested-With': 'OnlineShopping.WebApp'
        }

        response = requests.get(url, headers=headers, cookies=cookies, timeout=60)
        # if the response failed, raise an error
        if response.status_code != requests.codes.ok:
            response.raise_for_status()

        print(response.content)

        response_object = response.json()

        return response_object

    @staticmethod
    def create_price(response_object):
        """
        Return an instance of ProductPriceModel with appropriate details such as original price, sale price, date, if
        the product is on sale or not
        """

        # get original price and sale price
        original_price = response_object["price"]["originalPrice"]
        sale_price = response_object["price"]["salePrice"]

        product_size = response_object['size']['volumeSize']
        package_type = response_object['size']['packageType']

        product_quantity = product_size

        if package_type == 'each' and product_size is None:
            product_quantity = 'ea'

        if package_type == 'bunch' and product_size is None:
            product_quantity = 'bunch'

        # check if volume = "per kg" for fruits and vegetables and convert to 1kg
        if product_size == 'per kg':
            product_quantity = '1kg'

        if product_size == '1kg pack':
            product_quantity = '1kg'

        if product_size == '4 serve':
            product_quantity = '4pk'

        # if toiler paper is in name, it will have no quantity, so search in product name for quantity
        if 'toilet paper' in response_object['name'] and 'pk' in response_object['name']:
            try:
                product_quantity = re.search('([0-9]+)(pk)', response_object['name']).group()
            except AttributeError:
                print('Could not retrieve quantity from name')

        # check if product is on sale
        if sale_price < original_price:
            product_on_sale = True
        else:
            product_on_sale = False

        if original_price == 0:
            is_available = False
        else:
            is_available = True

        timezone = pytz.timezone('Pacific/Auckland')

        # set price with other details
        price = ProductPriceModel(datetime.now(timezone).date(), original_price, sale_price,
                                  product_on_sale, is_available, product_quantity)
        return price


c = CountdownPriceRetriever()

c.request_product_price("282769")

File: Spendy/PriceRetriever/PriceRetrievers/CountdownPriceRetriever.cs
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;
using PriceRetriever.Interfaces;

namespace PriceRetriever.PriceRetrievers;

public class CountdownPriceRetriever : IPriceRetriever
{
    private readonly HttpClient _client;

    public CountdownPriceRetriever(HttpClient client)
    {
        _client = client;
    }
    
    // countdown price model
    private class CountdownPriceResponseModel
    {   
        // price models for prices
        public class CountdownPriceModel
        {
            [JsonPropertyName("originalPrice")] public double? OriginalPrice { get; set; }
            [JsonPropertyName("salePrice")] public double? SalePrice { get; set; }
        }
        
        // size model for quantity
        public class CountdownSizeModel
        {
            [JsonPropertyName("volumeSize")] public string? VolumeSize { get; set; }
            [JsonPropertyName("packageType")] public string? PackageType { get; set; }
        }
        
        [JsonPropertyName("name")] public string? Name { get; set; }
        [JsonPropertyName("price")] public CountdownPriceModel? Price { get; set; }
        [JsonPropertyName("size")] public CountdownSizeModel? Size { get; set; }
    }
    
    public async Task<PriceModel> RetrievePrice(string storeProductCode)
    {

        var url = $"https://shop.countdown.co.nz/api/v1/products/{storeProductCode}";
        
        // _client.DefaultRequestVersion = new Version(2, 0);
        // add headers
        // _client.DefaultRequestHeaders.Accept.Clear();
        // _client.DefaultRequestHeaders.Accept.Add(
        //     new MediaTypeWithQualityHeaderValue("application/json"));
        // _client.DefaultRequestHeaders.Add("User-Agent",
            // "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:99.0) Gecko/20100101 Firefox/99.0");
        // client.DefaultRequestHeaders.Add("Accept-Encoding", "gzip, deflate, br");
        // _client.DefaultRequestHeaders.Add("ContentType", "application/json");
        // _client.DefaultRequestHeaders.Add("X-Requested-With", "OnlineShopping.WebApp");

        var response = await _client.GetStringAsync(url);
        
        // deserialize into CountdownPriceResponseModel
        var countdownPrice = JsonSerializer.Deserialize<CountdownPriceResponseModel>(response);

        // required price information
        var originalPrice = countdownPrice?.Price?.OriginalPrice;
        var salePrice = countdownPrice?.Price?.SalePrice;
        var isOnSale = salePrice < originalPrice;
        var isAvailable = originalPrice > 0;
        var priceDate = TimeZoneInfo.ConvertTime(DateTime.Now, TimeZoneInfo
            .FindSystemTimeZoneById("New Zealand Standard Time")).Date;
        var volumeSize = countdownPrice?.Size?.VolumeSize;
        var packageType = countdownPrice?.Size?.PackageType;
        var name = countdownPrice?.Name;
        
        // get quantity of the product and check for edge case quantities
        // check volume size first
        var priceQuantity = volumeSize switch
        {
            "per kg" => "1kg",
            "1kg pack" => "1kg",
            "4 serve" => "4pk",
            // if volume size contains no valid quantities (is null) check package type
            _ => packageType switch
            {
                "each" when volumeSize == null => "ea",
                "bunch" when volumeSize == null => "bunch",
                _ => volumeSize
            }
        };

        // check name for toilet paper, as toilet paper will not have any quantity info in response
        if (name != null && name.Contains("toilet paper") && name.Contains("pk"))
        {
            try
            {
                priceQuantity = Regex.Match(name, "([0-9]+)(pk)").ToString();
            }
            catch (Exception e)
            {
                Console.WriteLine($"The following error occured when attempting to get quantity from name: {e}");
                throw;
            }
        }
        
        // create price model
        var price = new PriceModel
        {
            OriginalPrice = originalPrice ?? 0,
            SalePrice = salePrice ?? 0,
            IsOnSale = isOnSale,
            IsAvailable = isAvailable,
            PriceDate = priceDate,
            PriceQuantity = priceQuantity ?? "none"
        };
        
        return price;
    }
}